File: App/food_store/lib/models/invoice.dart
import 'package:food_store/models/product/product.model.dart';

class DetailInvoice {
  final String id;
  final String productId;
  final ProductModel product;
  final int quantity;
  final int price;
  final String address;
  final String date;

  DetailInvoice({
    required this.id,
    required this.productId,
    required this.product,
    required this.quantity,
    required this.price,
    required this.address,
    required this.date,
  });

  factory DetailInvoice.fromJson(Map<String, dynamic> json) {
    return DetailInvoice(
      id: json['_id'],
      productId: json['productId'],
      product: ProductModel.fromJson(json['product']),
      quantity: json['quantity'],
      price: json['price'],
      address: json['address'],
      date: json['date'],
    );
  }
}

class Invoice {
  final String id;
  final String nameInvoice;
  final List<DetailInvoice> details;
  final String accountId;

  Invoice({
    required this.id,
    required this.nameInvoice,
    required this.details,
    required this.accountId,
  });

  factory Invoice.fromJson(Map<String, dynamic> json) {
    return Invoice(
      id: json['_id'],
      nameInvoice: json['nameInvoice'],
      details: (json['details'] as List)
          .map((detail) => DetailInvoice.fromJson(detail))
          .toList(),
      accountId: json['accountId'],
    );
  }
}

File: Api/features/models/invoice/invoice.js
const { Schema, model } = require('mongoose');

const DetailInvoiceSchema = new Schema({
    productId: { type: String, required: true },
    product: { type: Schema.Types.ObjectId, required: true, ref: 'Product' },
    quantity: { type: Number, required: true },
    price: { type: Number, required: true },
    address: { type: String, required: true },
    date: { type: String, required: true },
});
const DetailInvoiceModel = model('DetailInvoice', DetailInvoiceSchema);

const InvoiceSchema = new Schema({
    nameInvoice: { type: String, required: true },
    details: [{ type: Schema.Types.ObjectId, ref: 'DetailInvoice' }],
    accountId: { type: String, required: true },
    account: { type: Schema.Types.ObjectId, ref: 'Accounts' },
    status: { type: String, enum: ['active', 'cancelled', 'completed'], default: 'active' },
    cancelReason: { type: String },
    completedAt: { type: Date }
});

const InvoiceModel = model('Invoice', InvoiceSchema);
module.exports = { InvoiceModel, DetailInvoiceModel };